In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.7 MB 5.0 MB/s 
     |████████████████████████████████| 49 kB 4.2 MB/s 


In [2]:
from pyomo.environ import * 

In [3]:
import numpy as np

In [4]:
import pandas as pd

In [5]:
data=np.loadtxt("lab7_ex2_lp_ip_coef.txt",dtype='str')

In [6]:
data

array(['801,956,512,823,740,752,936,536,851,399,515,660,523,229,637,609,663,561,516,709,1060,405,564,692,524,992,745,860,638,801,777,757,722,785,321,943,469,541,673,716,686,1188,969,804,728,790,741,790,545,591,1091,707,475,808,750,707,556,573,562,1057,750,919,511,440,638,865,588,891,731,548,437,1015,878,1017,824,993,905,600,664,890,732,696,703,348,690,392,1090,1170,799,1191,561,840,555,918,698,687,735,572,926,599,469,486,973,676,817,504,388,685,596,878,426,862,695,542,746,694,885,524,694,935,715,810,978,443,318,621,805,511,1279,832,974,737,857,756,595,830,526,968,974,1279,770,980,435,881,858,964,454,1068,956,874,872,623,609,623,649,620,850,522,999,959,593,624,488,386,835,741,1045,815,809,860,472,949,972,874,951,568,890,633,622,509,788,496,1004,562,700,637,666,340,658,832,778,616,643,521,505,900,543,873,763,890,776,752,718,366,566,802,304,470,543,847,827,821,644,1012,845,702,624,942,963,510,428,710,986,663,928,541,933,1095,699,756,733,632,732,1073,552,913,514,1011,772,597,652,863,534,65

In [7]:
model=ConcreteModel()

In [8]:
data.shape

(6,)

In [9]:
len(data[0])

2040

In [10]:
obj_coef_c = data[0].split(',')

In [11]:
len(data[1].split(','))

501

In [12]:
constraint_coef=[]
for i in data[1:]:
  constraint_coef.append(i.split(','))

In [13]:
len(constraint_coef)

5

In [14]:
constr_rhs_b=[]
for i in constraint_coef:
  constr_rhs_b.append(i[500])

In [15]:
constr_rhs_b

['173725', '190019', '186926', '192737', '183645']

In [16]:
N=5
M=501
row_indices=np.arange(N)
col_indices=np.arange(M)

In [17]:
print(row_indices)

[0 1 2 3 4]


In [18]:
model.x = Var(col_indices[:-1], domain=Binary)

In [19]:
model.x

In [20]:
model.constraints = ConstraintList()

In [21]:
for i in row_indices:
    model.constraints.add(sum(int(constraint_coef[i][j])*model.x[j] for j in col_indices[:-1]) <= int(constr_rhs_b[i]))

In [22]:
model.objective = Objective(expr= sum(int(obj_coef_c[i])*model.x[i] for i in col_indices[:-1]), sense=maximize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :  500 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 15

In [23]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 159447 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [24]:
opt_cbc = SolverFactory('cbc')

In [25]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('Solver time:',result.solver.time)

Solver status: ok
Solver termination condition: optimal
Solver time: 73.35819220542908


In [26]:
print('Branch and BOund Statistics',result.solver.statistics)

Branch and BOund Statistics 
Branch and bound: 
  Number of bounded subproblems: 347044
  Number of created subproblems: 347044
Black box: 
  Number of iterations: 1407089



In [27]:
print(result)


Problem: 
- Name: unknown
  Lower bound: 295828.0
  Upper bound: 295828.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 500
  Number of binary variables: 500
  Number of integer variables: 500
  Number of nonzeros: 500
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 71.26
  Wallclock time: 73.29
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 347044
      Number of created subproblems: 347044
    Black box: 
      Number of iterations: 1407089
  Error rc: 0
  Time: 73.35819220542908
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [28]:
print('\nObjective = ', model.objective())


Objective =  295828.0


Q1: Optimal solution obtained is:295828.0 \\
Q2: the solver time is 77.41699266433716

Q3:
We are to change domain now.first we took domain binary now we take domain [0,1] interval.

In [29]:
model.x.domain=Reals
for i in col_indices[:-1]:
   model.x[i].setlb(0)
   model.x[i].setub(1)

In [30]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('Solver time:',result.solver.time)

Solver status: ok
Solver termination condition: optimal
Solver time: 0.043744564056396484


In [31]:
print(result)


Problem: 
- Name: unknown
  Lower bound: 295896.377
  Upper bound: 295896.377
  Number of objectives: 1
  Number of constraints: 6
  Number of variables: 501
  Number of nonzeros: 500
  Sense: maximize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.01
  Wallclock time: 0.01
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 18
  Error rc: 0
  Time: 0.043744564056396484
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [32]:
print('\nObjective = ', model.objective())


Objective =  295896.37697236


Q3:sol: when we changes restrictions on domain of variables we get the following value of objective funcion:295896.37697236

Q4:sol: solver time when domain is in [0,1] interval :0.030646324157714844
\\
Here solver time is less than the condition when domain was binary since it takes less time to check because when domain is binary it is restricted  to take only 0 and 1 value but in interval it can take any value in interval.

In [33]:
print('Branch and BOund Statistics',result.solver.statistics)

Branch and BOund Statistics 
Branch and bound: 
  Number of bounded subproblems: None
  Number of created subproblems: None
Black box: 
  Number of iterations: 18



 Q7 
 sol. for MILP:
 Branch and BOund Statistics \\
Branch and bound: \\ 
  Number of bounded subproblems: 347044 \\
  Number of created subproblems: 347044 \\
Black box:


  Number of iterations: 1407089 \\
for LP:Branch and BOund Statistics 

Branch and bound:

  Number of bounded subproblems: None \\
  Number of created subproblems: None \\
Black box: 
 
  Number of iterations: 18 \\
